// @sap/cds-runtime/lib/common/constants/draft.js
module.exports = {
  DRAFT_COLUMNS: ['IsActiveEntity', 'HasActiveEntity', 'HasDraftEntity', 'DraftAdministrativeData_DraftUUID'],
  DRAFT_COLUMNS_MAP: {
    IsActiveEntity: true,
    HasActiveEntity: true,
    HasDraftEntity: true,
    DraftAdministrativeData_DraftUUID: true,
    DraftUUID: true
  },
  DRAFT_COLUMNS_UNION: [
    'IsActiveEntity',
    'HasActiveEntity',
    'HasDraftEntity',
    'DraftAdministrativeData_DraftUUID',
    'SiblingEntity',
    'DraftAdministrativeData'
  ],
  DRAFT_COLUMNS_ADMIN: [
    'DraftUUID',
    'CreatedByUser',
    'InProcessByUser',
    'CreationDateTime',
    'LastChangeDateTime',
    'LastChangedByUser',
    'DraftIsProcessedByMe',
    'DraftIsCreatedByMe'
  ],
  DRAFT_COLUMNS_CQN: [
    { ref: ['IsActiveEntity'], cast: { type: 'cds.Boolean' } },
    { ref: ['HasActiveEntity'], cast: { type: 'cds.Boolean' } },
    { ref: ['HasDraftEntity'], cast: { type: 'cds.Boolean' } },
    { ref: ['DraftAdministrativeData_DraftUUID'] }
  ],
  DRAFT_COLUMNS_FOR_CQN_SELECT: [
    { val: true, as: 'IsActiveEntity', cast: { type: 'cds.Boolean' } },
    { val: false, as: 'HasActiveEntity', cast: { type: 'cds.Boolean' } },
    { val: false, as: 'HasDraftEntity', cast: { type: 'cds.Boolean' } }
  ]
}
